### PR TITLE
kernel: Add Reset component and capsule.

### PR DESCRIPTION
**Pull request by osenft**
This adds a HIL, implementation and capsule to execute a chip reset and to get the source of the last reset.

Depends on PR #237.

### CODE REVIEW

File: kernel/h1/src/pmu.rs
Comment (dann21) [inline kernel/h1/src/pmu.rs]: Where does this value come from?  (Maybe add a reference to the comment?)

File: kernel/h1/src/pmu.rs
Comment (dann21) [inline kernel/h1/src/pmu.rs:170]: Maybe add a comment here that this is the base-address of the PMU registers?
It probably seems obvious to you, but I did a double-take when I saw that this value was getting coerced into a pointer to a PMURegisters object (below).  "How can it be right to coerce an isize into a pointer?"

File: kernel/h1/src/pmu.rs
Comment (dann21) [inline kernel/h1/src/pmu.rs]: This may just be that I don't know rust at all, but I find this line to be confusing.  It uses one name `[rR]eset` for three different things (value, object and trait, if I'm reading it correctly).  Is that standard practice?  (Yes, I know, it's rust.  If it was hard to write, it should be hard to read.  :-) )

File: kernel/h1/src/pmu.rs
Comment (dann21) [inline kernel/h1/src/pmu.rs:296]: It seems ugly to me that you need to do a `memcpy()` (== `transmute()`) to coerce from a `*mut PMURegisters` to a `&mut PMURegisters`.  Isn't there a more principled way to do that?

Once again, I don't know the language.

File: kernel/h1/src/pmu.rs
Comment (dann21) [inline kernel/h1/src/pmu.rs:314]: The casual reader (me) has no idea at all what these instructions are doing.

Furthermore, a loop that spins, executing two instructions one after another forever seems wrong.  Presumably that's not what it's doing, because presumably the sequence causes the machine to reset, but I had to infer that.  There should at least be a comment that says that.

Finally, if the two-instruction sequence should always cause a reset, then this is not the way I would write it -- I would write it as something like `dsb`; `wfi`; `UNREACHABLE()`.

Either that or I have no clue what's going on here.

File: kernel/h1_syscalls/src/reset.rs
Comment (dann21) [inline kernel/h1_syscalls/src/reset.rs]: Is there no such thing as `UNREACHABLE()` to use here?

**Review by dann21 — COMMENTED** [review]
I think I understand most of what's going on here, but I think I had a sufficient number of questions that I'll wait on approval until you make clear the depth of my cluelessness.

File: kernel/h1/src/pmu.rs
Comment (osenft) [reply kernel/h1/src/pmu.rs:296]: Yeah, there are better (I think) ways that I used for the implementation in other drivers, e.g. look at "fuse.rs".

However, the pmu.rs already existed and I didn't want to refactor the whole file. I just followed the style that was used in the existing code in this file.

File: kernel/h1/src/pmu.rs
Comment (osenft) [reply kernel/h1/src/pmu.rs:314]: I actually got that sequence from some of our existing code for this chip. I agree it's a little confusing, but I don't think it's exciting enough. It certainly does what we want it to do ...

File: kernel/h1/src/pmu.rs
Comment (osenft) [reply kernel/h1/src/pmu.rs]: Good point. Done.

File: kernel/h1/src/pmu.rs
Comment (osenft) [reply kernel/h1/src/pmu.rs:170]: Done.

File: kernel/h1/src/pmu.rs
Comment (osenft) [reply kernel/h1/src/pmu.rs]: There are actually 5 different "reset" things:
* struct (Reset)
* trait (Reset)
* namespace (reset)
* method (reset)
* instance (RESET)

I renamed it to make it a little clearer:
* struct (ResetImpl)
* trait (Reset)
* namespace (reset)
* method (reset_chip)
* instance (RESET)

File: kernel/h1_syscalls/src/reset.rs
Comment (osenft) [reply kernel/h1_syscalls/src/reset.rs]: I updated the comment. The compiler would flag it if `self.reset.reset_chip()` changed its return type and actually returned.

File: kernel/h1/src/pmu.rs
Comment (dann21) [reply kernel/h1/src/pmu.rs]: :+1:

File: kernel/h1/src/pmu.rs
Comment (dann21) [reply kernel/h1/src/pmu.rs:296]: ok

File: kernel/h1_syscalls/src/reset.rs
Comment (dann21) [reply kernel/h1_syscalls/src/reset.rs]: :+1:

File: kernel/h1/src/pmu.rs
Comment (dann21) [reply kernel/h1/src/pmu.rs:314]: I guess the comment above the loop gives enough of a high-level view of what's going on, although I'd love to know why it needs to be in a loop.  Perhaps it's written that way to convince the compiler that it never returns?

**Review by dann21 — APPROVED** [review]
LGTM, modulo whatever is causing the build failure.

**Pull request merged**
